### PR TITLE
security: harden preprocess.py against FFmpeg SSRF, symlink TOCTOU & CSV injection (#191)

### LINKED ISSUE #191: Vulnerability Report: Critical Insecure FFmpeg Usage, Arbitrary File Write, and ML Pipeline Data Poisoning in VintageVoice
VULNERABILITY REPORT

Title: Critical Insecure FFmpeg Usage, Arbitrary File Write, and ML Pipeline Data Poisoning in VintageVoice
Date: June 17, 2026
Affected Repository: vintage-voice (specifically scripts/preprocess.py and run_pipeline.sh)
Severity: Critical (CVSS v3.1 Estimated: 9.0)
CWE IDs: CWE-918 (SSRF), CWE-59 (Symlink Attack), CWE-367 (TOCTOU), CWE-1236 (CSV Injection), CWE-507 (Data Poisoning)

1. EXECUTIVE SUMMARY

The vintage-voice repository contains a machine learning audio preprocessing pipeline (preprocess.py and run_pipeline.sh) designed to convert, split, and transcribe audio for F5-TTS training. 

A thorough security review reveals multiple critical vulnerabilities stemming from insecure external process execution (FFmpeg), unsafe file system operations, and a lack of input validation across pipeline stages. These flaws allow an attacker with write access to the input directory (or via a file upload endpoint) to achieve Server-Side Request Forgery (SSRF), Arbitrary File Read/Write, and Machine Learning Data Poisoning, ultimately compromising the host system and the integrity of the trained AI model.

2. VULNERABILITY DETAILS

2.1. Insecure FFmpeg Protocol Handling (SSRF & Arbitrary File Read)
Severity: High
CWE: CWE-918 (Server-Side Request Forgery)

Description:
The convert_to_wav, split_on_silence, and check_audio_quality functions invoke ffmpeg and ffprobe without restricting allowed protocols. FFmpeg supports numerous protocols (e.g., http, https, ftp, file) and performs content sniffing (magic bytes) regardless of the file extension. 

An attacker can craft a malicious file with a valid audio extension (e.g., .mp3) but containing an HLS playlist (.m3u8) payload. When FFmpeg processes this file, it will execute the URLs defined in the playlist.

Proof of Concept (PoC):
1. Create a malicious file in the input directory:
   echo -e "#EXTM3U\n#EXTINF:10.0,\nfile:///etc/passwd\n#EXT-X-ENDLIST" > /mnt/18tb/vintage_voice/malicious.mp3
2. Run the pipeline: bash run_pipeline.sh 1
3. Impact: FFmpeg will attempt to read /etc/passwd. By changing the payload to http://169.254.169.254/latest/meta-data/, the attacker can achieve SSRF to steal cloud instance credentials.

2.2. TOCTOU Race Condition & Symlink Attack (Arbitrary File Write)
Severity: Critical
CWE: CWE-367 (Time-of-check Time-of-use), CWE-59 (Improper Link Resolution)

Description:
In preprocess.py, the process_one_file function checks for file existence before writing:
wav_path = os.path.join(wav_dir, f"{stem}.wav")
if not os.path.exists(wav_path):
    ok = convert_to_wav(input_path, wav_path)

This is a classic TOCTOU vulnerability. If the output directory (/mnt/18tb/vintage_voice_processed/wav_full/) is on a shared filesystem or accessible by multiple users, an attacker can create a broken symlink at wav_path immediately after the os.path.exists() check but before ffmpeg writes the file. Because ffmpeg is executed with the -y (overwrite) flag, it will blindly write the WAV data to the symlink's target, allowing Arbitrary File Write.

Proof of Concept (PoC):
1. Create a broken symlink pointing to a sensitive file:
   ln -s /etc/cron.d/malicious_job /mnt/18tb/vintage_voice_processed/wav_full/target.wav
2. Trigger the processing of target.mp3.
3. Impact: ffmpeg overwrites /etc/cron.d/malicious_job with WAV audio data, corrupting the system file or allowing privilege escalation if the script runs as root.

2.3. Pipeline Data Poisoning & CSV Injection
Severity: High
CWE: CWE-1236 (CSV Injection), CWE-507 (Trojan Horse / Data Poisoning)

Description:
The run_pipeline.sh orchestrates a 5-step sequential pipeline. Step 1 (preprocess.py) generates a manifest.csv. Subsequent steps (Whisper transcription, F5-TTS dataset preparation, and Model Fine-tuning) blindly trust this manifest without validation. 

Furthermore, preprocess.py writes to manifest.csv using raw string formatting without sanitizing filenames:
f.write(f"{seg['path']}|{seg['duration']:.2f}|{Path(seg['path']).stem}\n")

An attacker can create files with newlines (\n) or pipe (|) characters in their names to inject malicious rows into the CSV, or inject paths to sensitive system files into the training dataset.

Proof of Concept (PoC):
1. Create a file with a newline in the name:
   touch $'/mnt/18tb/vintage_voice/etc/shadow_seg0000.wav\ninjected|999.00|pwned.mp3'
2. Impact: The manifest.csv structure is broken. If the attacker injects a path to a sensitive file, Step 2 (Whisper) will transcribe it, and Step 4 (Fine-tuning) will train the AI model on sensitive system data, resulting in a permanent Supply Chain / Data Poisoning attack.

3. IMPACT ASSESSMENT

- Confidentiality: Attackers can read arbitrary local files (e.g., /etc/shadow, SSH keys) or internal network metadata via SSRF.
- Integrity: Attackers can overwrite arbitrary system files via Symlink attacks. More critically, they can poison the machine learning training data, resulting in a compromised, backdoored, or degraded AI model.
- Availability: Malformed files can cause ffmpeg to hang, and the unmitigated ProcessPoolExecutor can lead to resource exhaustion (OOM) and Denial of Service.

4. REMEDIATION & MITIGATION

4.1. Secure FFmpeg Execution (Immediate Fix)
Restrict FFmpeg protocols and validate file magic bytes before processing.
Code snippet:
import magic

def validate_and_convert(input_path, output_path):
    # 1. Validate Magic Bytes
    mime = magic.from_file(input_path, mime=True)
    if not mime.startswith('audio/'):
        raise ValueError(f"Invalid file type: {mime}")

    # 2. Restrict Protocols
    cmd = [
        "ffmpeg", "-y", 
        "-protocol_whitelist", "file,pipe", # CRITICAL FIX
        "-i", input_path,
        # ... other args
    ]

4.2. Fix TOCTOU / Symlink Vulnerability
Replace os.path.exists() with atomic file creation using O_EXCL.
Code snippet:
import os

# Secure file creation
fd = os.open(wav_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
os.close(fd)
# Now safe to pass to ffmpeg

4.3. Pipeline Validation & Sanitization
- Sanitize Filenames: Strip or reject characters like \n, \r, and | from filenames before processing.
- Use Python's csv module: Replace manual string formatting with csv.writer to properly escape CSV fields.
- Inter-step Validation: Add a validation script between Step 1 and Step 2 in run_pipeline.sh to verify that all paths in manifest.csv exist, are within the allowed output directory, and match expected schemas.

4.4. Architectural Hardening (Long-term)
- Sandboxing: Execute FFmpeg inside a Docker container with --network=none, --read-only, and --security-opt=no-new-privileges.
- Least Privilege: Run the pipeline under a dedicated, unprivileged user with write permissions only to the specific output directories.

5. CONCLUSION

The current implementation of the vintage-voice preprocessing pipeline prioritizes functionality over security, relying heavily on the assumption that the execution environment and input data are trusted. By implementing the recommended fixes—specifically restricting FFmpeg protocols, securing file system operations, and validating pipeline data flow—the system can be secured against both local and remote exploitation vectors.

### PR DESCRIPTION
**Pull request by Vyacheslav-Tomashevskiy**
## What

Hardens `scripts/preprocess.py` against the three vulnerability classes reported in #191. The fix is self-contained in that one file and changes no public behaviour for legitimate audio.

## Fixes

**1. FFmpeg SSRF / arbitrary file read (CWE-918)**
- Every `ffmpeg`/`ffprobe` call now passes `-protocol_whitelist file,pipe`, so a crafted file can no longer make FFmpeg follow `http://169.254.169.254/...` (cloud metadata) or other network URLs.
- New `assert_safe_audio_input()` rejects files whose *content* is a playlist/markup payload (`#EXTM3U`, `#EXT-X`, `ffconcat`, `<?xml`, `<`) regardless of extension — this neutralises the disguised-`.mp3` HLS PoC that reads `file:///etc/passwd`.

**2. Symlink / TOCTOU arbitrary write (CWE-367, CWE-59)**
- `convert_to_wav` and `extract_segment` now write to a `mkstemp()` temp file in the destination directory and `os.replace()` it onto the final path. `os.replace` is atomic and overwrites the link itself, never its target — so a symlink planted at the output path can no longer redirect the write to `/etc/cron.d/...`.
- `process_one_file` additionally refuses to process when the output path is already a symlink.

**3. CSV injection / ML data poisoning (CWE-1236, CWE-507)**
- The manifest is written with `csv.writer(delimiter="|", quoting=csv.QUOTE_MINIMAL)` instead of raw f-string formatting, so filenames containing `|`, `\n` or `"` are properly quoted and cannot inject extra manifest rows.
- Derived output filenames are sanitised via `_safe_stem()`.
- Format stays compatible with the existing `csv.DictReader(f, delimiter="|")` consumer in `transcribe_cajun.py`.

## Verification

Tested end-to-end with real `ffmpeg`:
- Legitimate audio still converts and segments correctly (no regression).
- The disguised-playlist PoC from the report is rejected before reaching FFmpeg.
- The symlink-write PoC no longer overwrites the link's target (verified the target file is left untouched).
- A `|`/newline-poisoned filename produces exactly one correctly-quoted manifest row when re-parsed with `csv.DictReader`.

Closes #191.

### CODE REVIEW

**Comment by Vyacheslav-Tomashevskiy** [comment]
/claim

This PR fixes all three vulnerability classes reported in #191 (FFmpeg SSRF/arbitrary read, symlink TOCTOU arbitrary write, and CSV-injection data poisoning), self-contained in `scripts/preprocess.py` and verified end-to-end with real ffmpeg. Closes #191.

**Review by FakerHideInBush — CHANGES_REQUESTED** [review]
Solid hardening work overall — the FFMPEG_PROTOCOLS whitelist and atomic write via os.replace() are the right foundations. Three issues that should be addressed before merge:

**1. _ffmpeg_to_temp swallows all exceptions, masking security-critical failures**

The bare except Exception clause means PermissionError (tmpdir not writable), OSError, and failures from malicious input all collapse to return False. Callers cannot distinguish between 'bad audio file rejected' and 'filesystem write failed silently.' Worse: if os.unlink(tmp_path) raises inside the except block, that second exception is silently dropped, potentially leaking temp files. Consider re-raising non-conversion errors or at minimum logging them so audits can distinguish security events from operational failures.

**2. assert_safe_audio_input is not called from split_on_silence**

convert_to_wav now validates the header before invoking FFmpeg. But split_on_silence also feeds a wav_path argument directly to FFmpeg commands without calling assert_safe_audio_input. A caller who sources a WAV from an untrusted path and passes it to split_on_silence bypasses the header check entirely. Either add validation at the top of split_on_silence, or document explicitly that callers must validate before calling it.

**3. file in FFMPEG_PROTOCOLS still allows arbitrary local-file reads via playlist indirection if validation is bypassed**

FFMPEG_PROTOCOLS = "file,pipe" correctly blocks HTTP/RTMP SSRF, but file:// remains in the whitelist. An HLS concat playlist containing file:///etc/shadow segments, if fed to FFmpeg (via a gap in assert_safe_audio_input or a future call site that skips validation), would enable local file exfiltration. Consider whether pipe alone is sufficient for the audio pipeline, or add -allowed_extensions restrictions to further limit what FFmpeg will dereference via file://.

**Comment by Scottcjn** [comment]
Merged into #207 — your work was reconciled with the other two overlapping preprocess PRs into one coherent change (and a few integration bugs fixed via a tri-brain review). Credited to you via co-author trailer. Thanks @Vyacheslav-Tomashevskiy! 🙏
